File: domain/src/main/java/ch/threema/domain/models/MessageId.java
/*  _____ _
 * |_   _| |_  _ _ ___ ___ _ __  __ _
 *   | | | ' \| '_/ -_) -_) '  \/ _` |_
 *   |_| |_||_|_| \___\___|_|_|_\__,_(_)
 *
 * Threema for Android
 * Copyright (c) 2013-2022 Threema GmbH
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License, version 3,
 * as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

package ch.threema.domain.models;

import ch.threema.base.ThreemaException;
import ch.threema.base.utils.Utils;
import ch.threema.domain.protocol.csp.ProtocolDefines;

import java.io.Serializable;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.security.SecureRandom;
import java.util.Arrays;

/**
 * Wrapper class for message IDs (consisting of 8 bytes, chosen by the sender and not guaranteed
 * to be unique across multiple senders).
 */
public class MessageId implements Serializable {

	private final byte[] messageId;

	/**
	 * Create a new random MessageId.
	 */
	public MessageId() {
		messageId = new byte[ProtocolDefines.MESSAGE_ID_LEN];
		SecureRandom rnd = new SecureRandom();
		rnd.nextBytes(messageId);
	}

	/**
	 * Create a MessageId from an 8-byte array.
	 * @throws ThreemaException If the source array has the wrong length.
	 */
	public MessageId(byte[] messageId) throws ThreemaException {
		if (messageId.length != ProtocolDefines.MESSAGE_ID_LEN) {
			throw new ThreemaException("Invalid message ID length");
		}
		this.messageId = messageId;
	}

	/**
	 * Create a MessageId from an array, starting at the specified offset.
	 */
	public MessageId(byte[] data, int offset) {
		messageId = new byte[ProtocolDefines.MESSAGE_ID_LEN];
		System.arraycopy(data, offset, messageId, 0, ProtocolDefines.MESSAGE_ID_LEN);
	}

	/**
	 * Create a MessageId from an (unsigned) long in little-endian format
	 */
	public MessageId(long messageIdLong) {
		messageId = ByteBuffer.allocate(8).order(ByteOrder.LITTLE_ENDIAN).putLong(messageIdLong).array();
	}

	public byte[] getMessageId() {
		return messageId;
	}

	public long getMessageIdLong() {
		return ByteBuffer.wrap(messageId).order(ByteOrder.LITTLE_ENDIAN).getLong();
	}

	@Override
	public String toString() {
		return Utils.byteArrayToHexString(messageId);
	}

	@Override
	public boolean equals(Object obj) {
		if (obj == null)
			return false;
		if (obj == this)
			return true;
		if (obj.getClass() != getClass())
			return false;

		return Arrays.equals(messageId, ((MessageId)obj).messageId);
	}

	@Override
	public int hashCode() {
		/* message IDs are usually random, so just taking the first four bytes is fine */
		return messageId[0] << 24 | (messageId[1] & 0xFF) << 16 | (messageId[2] & 0xFF) << 8 | (messageId[3] & 0xFF);
	}
}

File: domain/src/main/java/ch/threema/domain/models/MessageId.kt
/*  _____ _
 * |_   _| |_  _ _ ___ ___ _ __  __ _
 *   | | | ' \| '_/ -_) -_) '  \/ _` |_
 *   |_| |_||_|_| \___\___|_|_|_\__,_(_)
 *
 * Threema for Android
 * Copyright (c) 2013-2025 Threema GmbH
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License, version 3,
 * as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

package ch.threema.domain.models

import ch.threema.base.ThreemaException
import ch.threema.common.generateRandomBytes
import ch.threema.common.secureRandom
import ch.threema.common.toByteArray
import ch.threema.common.toLong
import ch.threema.domain.protocol.csp.ProtocolDefines
import ch.threema.domain.types.MessageIdString
import java.io.Serializable
import java.nio.ByteOrder
import java.security.SecureRandom

/**
 * Wrapper class for message IDs (consisting of [ProtocolDefines.MESSAGE_ID_LEN] bytes, chosen by the sender and not guaranteed
 * to be unique across multiple senders).
 *
 * @throws IllegalArgumentException if the [messageId] is of the wrong length
 */
class MessageId(
    val messageId: ByteArray,
) : Serializable {

    init {
        require(messageId.size == ProtocolDefines.MESSAGE_ID_LEN)
    }

    /**
     * Create a MessageId from an array, starting at the specified [offset].
     *
     * @throws IndexOutOfBoundsException if fewer than [ProtocolDefines.MESSAGE_ID_LEN] bytes are available in [data] at [offset].
     */
    constructor(
        data: ByteArray,
        offset: Int,
    ) : this(data.copyOfRange(offset, offset + ProtocolDefines.MESSAGE_ID_LEN))

    /**
     * Create a MessageId from an (unsigned) long in little-endian format
     */
    constructor(
        messageIdLong: Long,
    ) : this(messageIdLong.toByteArray(order = ByteOrder.LITTLE_ENDIAN))

    val messageIdLong: Long
        get() = messageId.toLong(order = ByteOrder.LITTLE_ENDIAN)

    override fun toString(): MessageIdString = messageId.toHexString()

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false
        other as MessageId
        return messageId.contentEquals(other.messageId)
    }

    override fun hashCode(): Int {
        // message IDs are usually random, so just taking the first four bytes is fine
        return (messageId[0].toInt() shl 24) or
            ((messageId[1].toInt() and 0xFF) shl 16) or
            ((messageId[2].toInt() and 0xFF) shl 8) or
            (messageId[3].toInt() and 0xFF)
    }

    companion object {
        /**
         * Create a MessageId from a String
         *
         * @throws ThreemaException If the message id is `null` or has an invalid length
         */
        @JvmStatic
        fun fromString(messageId: MessageIdString?): MessageId {
            if (messageId == null) {
                throw ThreemaException("Message id is null")
            }
            try {
                return MessageId(messageId.hexToByteArray())
            } catch (e: IllegalArgumentException) {
                throw ThreemaException("Message id is invalid", e)
            }
        }

        /**
         * Create a new random MessageId.
         */
        @JvmStatic
        @JvmOverloads
        fun random(random: SecureRandom = secureRandom()): MessageId =
            MessageId(random.generateRandomBytes(ProtocolDefines.MESSAGE_ID_LEN))
    }
}
